### PR TITLE
Remove deprecated 'terminate_on_nan' argument from Trainer

### LINKED ISSUE #12521: Remove deprecated code after the 1.6 release
## Proposed refactor

Remove deprecated code after the 1.6 release.

**NOTE**: Please pick up a _single_ item from the list (by commenting here in the issue) - and if there are no conflicts - we will happily assign you and put your name in front of the item in the list.

Please note that unless mentioned, the classes are importable from `pytorch_lightning`, example: `from pytorch_lightning import Trainer`.

* [x] `LightningModule.summarize` -> #12559
* [x] `pytorch_lightning.core.memory.LayerSummary` -> #12593
* [x] `pytorch_lightning.core.memory.ModelSummary` -> #12593
* [x] `pytorch_lightning.core.memory.get_gpu_memory_map` -> #12644
* [x] `pytorch_lightning.core.memory.get_memory_profile` -> #12659
* [x] `LightningModule.model_size` -> #12641
* [x] `LightningDataModule.train_transforms` -> #12662
* [x] `LightningDataModule.val_transforms` -> #12763
* [x] `LightningDataModule.test_transforms` -> #12773
* [x] `LightningDataModule.size` -> #12780
* [x] `LightningDataModule.dims` and `LightningDataModule(dims=...)` -> #12780
* [x] `LightningModule.get_progress_bar_dict` -> #12839 
* [x] `Trainer.progress_bar_dict` -> #12839 
* [x] `Trainer(prepare_data_per_node=...)` -> #12536
* [x] `Trainer(stochastic_weight_avg=...)` -> #12535
* [x] `Trainer(terminate_on_nan=...)` and `Trainer.terminate_on_nan` -> #12553 
* [x] `LightningModule.on_{train,val,test,predict}_dataloader` -> #13033
* [x] `pytorch_lightning.loggers.TestTubeLogger` -> #12859
* [x] `pytorch_lightning.Callback.on_keyboard_interrupt` -> #13438
* [x] `Trainer(process_position=...)` -> #13071
* [x] `Trainer(flush_logs_every_n_steps=...)` -> #13074
* [x] `LightningModule.add_to_queue` -> @shenoynikhil
* [x] `LightningModule.get_from_queue` -> @shenoynikhil
* [x] `Trainer(progress_bar_refresh_rate=...)` -> #12514
* [x] `LightningLoggerBase.close` and `pytorch_lightning.loggers.LoggerCollection.close` -> #13149
* [x] `pytorch_lightning.distributed.dist.LightningDistributed` #13549
* [x] `Trainer(checkpoint_callback=...)` -> #13027
* [x]  Passing `dataloader_idx` to `on_train_batch_start` of `pytorch_lightning.Callback` and `LightningModule`  -> #12769
* [x] `LightningModule.on_post_move_to_device` #13548
* [x] `pytorch_lightning.core.decorators.parameter_validation` #13514
* [x] `Trainer(accelerator="ddp_spawn")` #12696
* [x] `Trainer(plugins="ddp_spawn")` #12700
* [x] `Trainer(weights_summary="full")`, `Trainer(weights_summary=None)`, `Trainer.weights_summary` -> #13070
* [x] `Trainer(log_gpu_memory=...)` -> #12657
* [x] `Trainer.slurm_job_id` #13459
* [x] `pytorch_lightning.callbacks.gpu_stats.GPUStatsMonitor` -> #12554
* [x] `pytorch_lightning.callbacks.gpu_stats.XLAStatsMonitor` -> #12688
* [x] `pytorch_lightning.callbacks.progress.ProgressBar` -> #12658
* [x] `Trainer(max_steps=None)` and `Trainer.fit_loop.max_steps = None` #13591
* [x] `pytorch_lightning.callbacks.lr_monitor.LearningRateMonitor.lr_sch_names` -> #13353
* [x] `KubeflowEnvironment.is_using_kubeflow, LSFEnvironment.is_using_lsf, TorchElasticEnvironment.is_using_torchelastic` #13458
* [x] `pytorch_lightning.overrides.distributed.IndexBatchSamplerWrapper.batch_indices` #13565
* [x] `pytorch_lightning.strategies.SingleDeviceStrategy.post_dispatch` #13461
* [x] `pytorch_lightning.trainer.connectors.logger_connector.logger_connector.LoggerConnector.gpu_metrics`

Feel free to cross-check from the [test file](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/deprecated_api/test_remove_1-7.py) to ensure that the relevant test fails now (since it's no more deprecated and instead removed).

### Pitch
All the deprecated features we have are tested here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/deprecated_api/test_remove_1-7.py

If you are interested in taking care of one item, post a comment here asking to take it. This avoids multiple people working on the same thing.

### Additional context
See pull requests linked in #10312 for examples on how to contribute :)
Or a recent pull request #12514.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @awaelchli @rohitgr7 @krshrimali

### PR DESCRIPTION
**Pull request by karthikrangasai**
## What does this PR do?

Part of #12521

### Does your PR introduce any breaking changes? If yes, please list them.
Remove deprecated 'terminate_on_nan' argument from Trainer.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/optimization/optimizer_loop.py [hunk 1]
```diff
@@ -310,10 +308,6 @@ def _make_backward_fn(self, optimizer: Optimizer, opt_idx: int) -> Optional[Call
         def backward_fn(loss: Tensor) -> None:
             self.trainer._call_strategy_hook("backward", loss, optimizer, opt_idx)
 
-            # check if model weights are nan
-            if self.trainer._terminate_on_nan:
-                detect_nan_parameters(self.trainer.lightning_module)
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimization/optimizer_loop.py:315]: added a test for this utility since it no longer gets used otherwise :)

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -395,16 +394,6 @@ def __init__(
...
 
-            terminate_on_nan: If set to True, will terminate training (by raising a `ValueError`) at the
-                end of each training batch, if any of the parameters or the loss are NaN or +/-inf.
-
-                .. deprecated:: v1.5
-                    Trainer argument ``terminate_on_nan`` was deprecated in v1.5 and will be removed in 1.7.
-                    Please use ``detect_anomaly`` instead.
-
-            detect_anomaly: Enable anomaly detection for the autograd engine.
-                Default: ``False``.
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:406]: removed this, since it was duplicated

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -904,72 +904,12 @@ def validation_epoch_end(self, *args, **kwargs):
...
-
-
 @pytest.mark.parametrize("track_grad_norm", [0, torch.tensor(1), "nan"])
 def test_invalid_track_grad_norm(tmpdir, track_grad_norm):
     with pytest.raises(MisconfigurationException, match="`track_grad_norm` must be a positive number or 'inf'"):
         Trainer(default_root_dir=tmpdir, track_grad_norm=track_grad_norm)
 
 
-@mock.patch("torch.Tensor.backward")
-def test_nan_params_detection(backward_mock, tmpdir):
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:949]: removed this test, because with detect_anomaly we don't stop training anymore and instead prints a warning (a test for this exists already)

**Review by awaelchli — APPROVED** [review]
great work!

**Review by rohitgr7 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]
